<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>GenomicDiversity.jl · GenomicDiversity</title><meta name="title" content="GenomicDiversity.jl · GenomicDiversity"/><meta property="og:title" content="GenomicDiversity.jl · GenomicDiversity"/><meta property="twitter:title" content="GenomicDiversity.jl · GenomicDiversity"/><meta name="description" content="Documentation for GenomicDiversity."/><meta property="og:description" content="Documentation for GenomicDiversity."/><meta property="twitter:description" content="Documentation for GenomicDiversity."/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href><img src="assets/logo.jpg" alt="GenomicDiversity logo"/></a><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>GenomicDiversity.jl</a><ul class="internal"><li><a class="tocitem" href="#Functions"><span>Functions</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>GenomicDiversity.jl</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>GenomicDiversity.jl</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/darreni/GenomicDiversity.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/darreni/GenomicDiversity.jl/blob/main/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="GenomicDiversity.jl"><a class="docs-heading-anchor" href="#GenomicDiversity.jl">GenomicDiversity.jl</a><a id="GenomicDiversity.jl-1"></a><a class="docs-heading-anchor-permalink" href="#GenomicDiversity.jl" title="Permalink"></a></h1><p>For an overview of this package with an example of its use, see <a href="https://github.com/darreni/GenomicDiversity.jl">here</a>.</p><p>Below are descriptions of all functions in the package.</p><h2 id="Functions"><a class="docs-heading-anchor" href="#Functions">Functions</a><a id="Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.GenoData" href="#GenomicDiversity.GenoData"><code>GenomicDiversity.GenoData</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">GenoData(indInfo::DataFrame, genotypes::Matrix, positions::DataFrame)</code></pre><p>Constructs a <code>GenoData</code> struct, which stores metadata of individuals, a genotype matrix of individuals and SNPS, and genomic position info.</p><p>​# Arguments / keywords</p><ul><li><code>indInfo</code>: A DataFrame containing metadata.</li><li><code>genotypes</code>: A Matrix containing genotypes for individuals (rows) and loci (columns).</li><li><code>positions</code>: A 2-column DataFrame containing genomic location info of SNPs.</li></ul><p><strong>Notes</strong></p><p>You can access each component of this data structure by appending &quot;.indInfo&quot; or &quot;.genotypes&quot; or &quot;positions&quot; to the name of the GenoData object.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L75-L87">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.chooseChrRegion-Tuple{Any, Any}" href="#GenomicDiversity.chooseChrRegion-Tuple{Any, Any}"><code>GenomicDiversity.chooseChrRegion</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">chooseChrRegion(pos, chr; positionMin=1, positionMax=NaN)</code></pre><p>Designate a specific region of a chromosome for subsequent analysis.</p><p>​# Arguments</p><ul><li><code>pos</code>: A matrix containing genomic location of each locus; must have a <code>chrom</code> column (name of scaffold) and a <code>position</code> column (numerical position on scaffold).</li><li><code>chr</code>: The name of the chosen scaffold.</li><li><code>positionMin</code>: Optional (default 1); the starting location of the chosen region.</li><li><code>positionMax</code>: Optional (default the end of the scaffold); the ending location of the chosen region.</li></ul><p><strong>Notes</strong></p><p>Returns a tuple containing:</p><ul><li>the scaffold name</li><li>the minimum position</li><li>the maximum position</li><li>a string describing the region (e.g. <code>chr Z: 20000 to 1000000</code>)</li></ul><p>This function does not actually filter the genotype matrix–that can be done by providing the output of this to functions such as <code>plotGenotypeByIndividual()</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L676-L695">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.chooseChrRegion-Tuple{GenoData, Any}" href="#GenomicDiversity.chooseChrRegion-Tuple{GenoData, Any}"><code>GenomicDiversity.chooseChrRegion</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">chooseChrRegion(myGenoData::GenoData, chr; positionMin=1, positionMax=NaN)</code></pre><p>Call chooseChrRegion using a <code>GenoData</code> object.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L709-L713">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.filterInds-Tuple{GenoData, Any}" href="#GenomicDiversity.filterInds-Tuple{GenoData, Any}"><code>GenomicDiversity.filterInds</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">filterInds(myGenoData::GenoData, selection; direction = &quot;out&quot;)</code></pre><p>Filter the individuals indicated by <code>selection</code>, and store result in a new <code>GenoData</code> object.  The default is to filter individuals out (remove them); set direction = &quot;in&quot; to filter them in (and remove others).</p><p>​# Arguments</p><ul><li><code>myGenoData</code>: A <code>GenoData</code> object.</li><li><code>selection</code>: A vector of Boolean values, or a BitVector.</li><li><code>direction</code>: (optional) The direction to filter (default is &quot;out&quot;)</li></ul><p><strong>Notes</strong></p><p>Returns a <code>GenoData</code> object, containing:</p><ul><li><code>indInfo</code>: A DataFrame containing metadata.</li><li><code>genotypes</code>: A Matrix containing genotypes for individuals (rows) and loci (columns).</li><li><code>positions</code>: A 2-column DataFrame containing genomic location info of SNPs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L150-L166">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.filterLoci-Tuple{GenoData, Any}" href="#GenomicDiversity.filterLoci-Tuple{GenoData, Any}"><code>GenomicDiversity.filterLoci</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">filterLoci(myGenoData::GenoData, selection; direction = &quot;out&quot;)</code></pre><p>Filter the loci indicated by <code>selection</code>, and store result in a new <code>GenoData</code> object.  The default is to filter loci out (remove them); set direction = &quot;in&quot; to filter them in (and remove others).</p><p>​# Arguments</p><ul><li><code>myGenoData</code>: A <code>GenoData</code> object.</li><li><code>selection</code>: A vector of Boolean values, or a BitVector.</li><li><code>direction</code>: (optional) The direction to filter (default is &quot;out&quot;)</li></ul><p><strong>Notes</strong></p><p>Returns a <code>GenoData</code> object, containing:</p><ul><li><code>indInfo</code>: A DataFrame containing metadata.</li><li><code>genotypes</code>: A Matrix containing genotypes for individuals (rows) and loci (columns).</li><li><code>positions</code>: A 2-column DataFrame containing genomic location info of loci.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L203-L219">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.filterNamedInds-Tuple{GenoData, Any}" href="#GenomicDiversity.filterNamedInds-Tuple{GenoData, Any}"><code>GenomicDiversity.filterNamedInds</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">filterNamedInds(myGenoData::GenoData, inds; direction = &quot;out&quot;)</code></pre><p>Filter out the individuals indicated by <code>selection</code>, and store result in a new <code>GenoData</code> object. </p><p>​# Arguments</p><ul><li><code>myGenoData</code>: A <code>GenoData</code> object.</li><li><code>inds</code>: A collection of names (strings) to filter out.</li><li><code>direction</code>: (optional) The direction to filter (default is &quot;out&quot;)</li></ul><p><strong>Notes</strong></p><p>Returns a <code>GenoData</code> object, containing:</p><ul><li><code>indInfo</code>: A DataFrame containing metadata.</li><li><code>genotypes</code>: A Matrix containing genotypes for individuals (rows) and loci (columns).</li><li><code>positions</code>: A 2-column DataFrame containing genomic location info of loci.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L181-L196">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.getDxy-Tuple{Any, Any}" href="#GenomicDiversity.getDxy-Tuple{Any, Any}"><code>GenomicDiversity.getDxy</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getDxy(freqs, groupsToCalc)</code></pre><p>Calculate per-locus Dxy (pi_between) for each pair of populations.</p><p><strong>Arguments</strong></p><ul><li><code>freqs</code>: Matrix containing alternate allele frequencies for each group (row) and SNP (column).</li><li><code>groupsToCalc</code>: Vector containing the names of the groups corresponding to each row of <code>freqs</code>.</li></ul><p><strong>Notes</strong></p><p>Returns a tuple containing: </p><ul><li>Matrix of Dxy values for each comparison (row) and locus (column).</li><li>Vector of comparison names.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L364-L377">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.getFreqsAndSampleSizes-Tuple{Any, Any, Any}" href="#GenomicDiversity.getFreqsAndSampleSizes-Tuple{Any, Any, Any}"><code>GenomicDiversity.getFreqsAndSampleSizes</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getFreqsAndSampleSizes(genotypes, indGroup, groupsToCalc)</code></pre><p>Calculate allele frequencies and sample sizes for each group and SNP.</p><p>​# Arguments</p><ul><li><code>genotypes</code>: The genotype matrix, where rows are individuals and columns are loci, with genotype codes 0,1,2 meaning homozygous reference, heterozygote, homozygous alternate, and missing genotypes can be either -1 or <code>missing</code>.</li><li><code>indGroup</code>: A vector providing the group name each individual belongs to.</li><li><code>groupsToCalc</code>: A list of group names to include in calculations.</li></ul><p><strong>Notes</strong></p><p>Returns a tuple containing 1) a matrix of frequencies, and 2) a matrix of samples sizes (in both, rows are groups and columns are loci). </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L266-L278">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.getFreqsAndSampleSizes-Tuple{GenoData, Any}" href="#GenomicDiversity.getFreqsAndSampleSizes-Tuple{GenoData, Any}"><code>GenomicDiversity.getFreqsAndSampleSizes</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getFreqsAndSampleSizes(myGenoData::GenoData, groupsToCalc)</code></pre><p>Call getFreqsAndSampleSizes() using a <code>GenoData</code> object as input.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L296-L300">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.getFst-Tuple{Any, Any, Any}" href="#GenomicDiversity.getFst-Tuple{Any, Any, Any}"><code>GenomicDiversity.getFst</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getFst(freqs, sampleSizes, groupsToCalc; among=false)</code></pre><p>Calculate Fst for each locus, between pairs of groups and (optionally) among all groups. </p><p>Fst is calculated according to Weir &amp; Cockerham (1984), including their correction for sample size and number of groups.</p><p>​# Arguments</p><ul><li><code>freqs</code>: Matrix containing alternate allele frequencies for each group (row) and SNP (column).</li><li><code>sampleSizes</code>: Matrix containing sample sizes for each group (row) and SNP (column).</li><li><code>groupsToCalc</code>: The list of group names.</li><li><code>among</code>: Optional argument–set to &quot;true&quot; to also calculate Fst among all groups.</li></ul><p><strong>Notes</strong></p><p>Returns a tuple containing: </p><ul><li>Matrix of Fst values for each comparison (row) and locus (column).</li><li>Matrix of numerator values from the Fst calculation for each comparison (row) and locus (column).</li><li>Matrix of denominator values from the Fst calculation for each comparison (row) and locus (column).</li><li>Vector of comparison names.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L411-L430">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.getPairwiseNames-Tuple{Any}" href="#GenomicDiversity.getPairwiseNames-Tuple{Any}"><code>GenomicDiversity.getPairwiseNames</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getPairwiseNames(groupsToCalc)::Vector{String}</code></pre><p>Using a list of names, return a list of paired names in format &quot;name1_name2&quot;.</p><p>​# Arguments</p><ul><li><code>groupsToCalc</code>: The list of names.</li></ul><p><strong>Notes</strong></p><p>Returns a vector of paired names.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L341-L351">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.getRegionDxy-Tuple{Any}" href="#GenomicDiversity.getRegionDxy-Tuple{Any}"><code>GenomicDiversity.getRegionDxy</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getRegionDxy(siteDxyMatrix)</code></pre><p>Given a matrix representing per-population-pair (rows) per-site (columns) Dxy values (within-group nucleotide diversity), produce a vector containing the mean Dxy for each population pair.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L393-L397">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.getRegionPi-Tuple{Any}" href="#GenomicDiversity.getRegionPi-Tuple{Any}"><code>GenomicDiversity.getRegionPi</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getRegionPi(sitePiMatrix)</code></pre><p>Given a matrix representing per-group (rows) per-site (columns) Pi values (within-group nucleotide diversity), produce a vector containing the mean pi for each group.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L324-L328">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.getRollingMean-Tuple{Any, Any}" href="#GenomicDiversity.getRollingMean-Tuple{Any, Any}"><code>GenomicDiversity.getRollingMean</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getRollingMean(inputVector, windowSize)</code></pre><p>Calculate non-overlapping windowed arithmetic average (i.e., means) from a vector of numeric values.</p><p>​# Arguments</p><ul><li><code>inputVector</code>: A vector of numeric values (can include NaN).</li><li><code>windowSize</code>: An integer indicating the window number of values to include in each mean.</li></ul><p><strong>Notes</strong></p><ul><li>Returns a vector of windowed means. </li><li>In the calculations, NaN values are ignored.</li><li>If all values in a window are NaN, then the mean is NaN.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L1230-L1243">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.getSitePi-Tuple{Any, Any}" href="#GenomicDiversity.getSitePi-Tuple{Any, Any}"><code>GenomicDiversity.getSitePi</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getSitePi(freqs, sampleSizes)</code></pre><p>Calculate non-biased nucleotide diversity (pi) at each site for each population, with correction for sample size, using allele freqs and sample size as input.</p><p>​# Arguments</p><ul><li><code>freqs</code>: Matrix containing alternate allele frequencies for each group (row) and SNP (column).</li><li><code>sampleSizes</code>: Matrix containing sample sizes for each group (row) and SNP (column).</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L307-L315">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.getWindowBoundaries-Tuple{Any, Any}" href="#GenomicDiversity.getWindowBoundaries-Tuple{Any, Any}"><code>GenomicDiversity.getWindowBoundaries</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getWindowBoundaries(lociVector, windowSize)</code></pre><p>Determine the start and end scaffold positions for each window along a single scaffold. </p><p>​# Arguments</p><ul><li><code>lociVector</code>: Vector of included locus positions along a scaffold.</li><li><code>windowSize</code>: Size of each window (in number of loci).</li></ul><p><strong>Notes</strong></p><ul><li>returns a Matrix, where rows represent windows and columns are starting location and ending location of each window along the scaffold. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L1470-L1481">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.getWindowedFst-NTuple{4, Any}" href="#GenomicDiversity.getWindowedFst-NTuple{4, Any}"><code>GenomicDiversity.getWindowedFst</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getWindowedFst(FstNumerator, FstDenominator, pos, windowSize)</code></pre><p>Calculate windowed Fst along chromosomal regions, uses output from each site&#39;s Fst (produced by the getFst() function). </p><p>This is calculated according to Weir&amp;Cockerham1984 (with sample size and pop number correction), calculated as windowed numerator over windowed denominator, in whole windows starting on left side of the region.</p><p>​# Arguments</p><ul><li><code>FstNumerator</code>: Produced by the getFst() function, this is a vector of the Fst numerator values for each locus. </li><li><code>FstDenominator</code>: Produced by the getFst() function, this is a vector of the Fst denominator values for each locus. </li><li><code>pos</code>: Matrix providing genomic location of each locus; there must be a <code>position</code> column.</li><li><code>windowSize</code>: An integer indicating the window number of values to include in each mean.</li></ul><p><strong>Notes</strong></p><ul><li>Returns a tuple of two vectors: rolling mean position, and rolling mean Fst.</li><li>The calculations ignore NaN values (which would be in both numerator and denominator).</li><li>Returns NaN for a window if there are no real values in the window. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L1261-L1278">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.getWindowedIndHet-Tuple{Any, Any, Any}" href="#GenomicDiversity.getWindowedIndHet-Tuple{Any, Any, Any}"><code>GenomicDiversity.getWindowedIndHet</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getWindowedIndHet(genotypes, pos, windowSize)</code></pre><p>Calculate windowed heterozygosity for each individual, across a genomic region.</p><p>​# Arguments</p><ul><li><code>genotypes</code>: The genotype matrix, where rows are individuals and columns are loci, with genotype codes 0,1,2 meaning homozygous reference, heterozygote, homozygous alternate, and missing genotypes can be either -1 or <code>missing</code>.</li><li><code>pos</code>: A matrix containing genomic location of each locus; must have a <code>position</code> column (numerical position on scaffold). </li><li><code>windowSize</code>: An integer indicating the number of loci to include in each mean.</li></ul><p><strong>Notes</strong></p><p>Returns a tuple containing 1) a vector of windowed mean positions, and 2) a matrix of windowed heterozgyosity values (rows are individuals and columns are loci). </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L1405-L1417">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.getWindowedViSHet-Tuple{Any}" href="#GenomicDiversity.getWindowedViSHet-Tuple{Any}"><code>GenomicDiversity.getWindowedViSHet</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getWindowedViSHet(windowedIndHet_standardized)</code></pre><p>Calculate the windowed Variance in Standardized Heterozygosity (ViSHet). </p><p>​# Arguments</p><ul><li><code>windowedIndHet_standardized</code>: The matrix of standardized individual windowed heterozygosity values (rows are individuals; columns represent windowed genomic regions)</li></ul><p><strong>Notes</strong></p><p>Returns a vector of variances in standardized individual heterozygosities (each value corresponds to a different windowed genomic region).  </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L1454-L1464">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.greet_GenomicDiversity-Tuple{}" href="#GenomicDiversity.greet_GenomicDiversity-Tuple{}"><code>GenomicDiversity.greet_GenomicDiversity</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">greet_GenomicDiversity()</code></pre><p>A simple function for testing that the module has loaded.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L63-L67">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.imputeMissingGenoData-Tuple{GenoData}" href="#GenomicDiversity.imputeMissingGenoData-Tuple{GenoData}"><code>GenomicDiversity.imputeMissingGenoData</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">imputeMissingGenotypes(myGenoData::GenoData; method = &quot;SVD&quot;)</code></pre><p>When applied to a GenoData object, imputes the data in the genotypes object, and returns a new GenoData object.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L255-L259">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.imputeMissingGenoData-Tuple{Matrix}" href="#GenomicDiversity.imputeMissingGenoData-Tuple{Matrix}"><code>GenomicDiversity.imputeMissingGenoData</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">imputeMissingGenotypes(genotypeMatrix::Matrix; method = &quot;SVD&quot;)</code></pre><p>Imputes missing values in the genotype matrix.</p><p>​# Arguments</p><ul><li><code>genotypeMatrix</code>: Matrix of genotypes, where missing data are indicated with either <code>-1</code> or <code>missing</code>.</li><li><code>method</code>: The method used for imputing. Presently &quot;SVD&quot; is the only option (and default).  </li></ul><p><strong>Notes</strong></p><p>Returns the genotype matrix with filled-in missing values, of type Float32. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L232-L243">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.limitIndsToPlot-NTuple{4, Any}" href="#GenomicDiversity.limitIndsToPlot-NTuple{4, Any}"><code>GenomicDiversity.limitIndsToPlot</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">limitIndsToPlot(plotGroups, numIndsToPlot, genoData, indMetadata)</code></pre><p>Select a subset of individuals, based on numbers within each group, for subsequent analysis.</p><p>​# Arguments</p><ul><li><code>plotGroups</code>: Vector of names of groups to include.</li><li><code>numIndsToPlot</code>: Vector of maximum number of individuals to include from each group (order as in <code>plotGroups</code>).</li><li><code>genoData</code>: Matrix of genotypes (rows are individuals, columns are loci).</li><li><code>indMetadata</code>: Matrix containing metadata for individuals (make sure there is an <code>Fst_group</code> column).</li><li><code>sortByMissing</code>: Set to true to choose individuals with the least missing data among these loci.</li></ul><p><strong>Notes</strong></p><p>Returns a tuple containing:</p><ul><li>Matrix of genotypes for included individuals.</li><li>Matrix of metadata for included individuals.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L486-L502">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.loadGenoData-NTuple{4, Any}" href="#GenomicDiversity.loadGenoData-NTuple{4, Any}"><code>GenomicDiversity.loadGenoData</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">loadGenoData(metadataFile, individualsFile, genotypesFile, positionsFile)</code></pre><p>Imports data from four files and builds a <code>GenoData</code> object. </p><p>​# Arguments</p><ul><li><code>metadataFile</code>: The path and name of the metadata file.</li><li><code>individualsFile</code>: The path and name of the file providing names of the individuals (corresponding to rows of the genotype matrix).</li><li><code>genotypesFile</code>: The path and name of the file containing the matrix of genotypes.</li><li><code>positionsFile</code>: The path and name of the file containing the genomic position info of the loci corresponding to columns of the genotype matrix.</li></ul><p><strong>Notes</strong></p><p>Returns a <code>GenoData</code> object, containing:</p><ul><li><code>indInfo</code>: A DataFrame containing metadata.</li><li><code>genotypes</code>: A Matrix containing genotypes for individuals (rows) and loci (columns).</li><li><code>positions</code>: A 2-column DataFrame containing genomic location info of SNPs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L104-L120">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.plotGenomeFst-NTuple{6, Any}" href="#GenomicDiversity.plotGenomeFst-NTuple{6, Any}"><code>GenomicDiversity.plotGenomeFst</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plotGenomeFst(scaffolds_to_plot, 
                windowed_Fst_all,
                pairwiseNamesFst,
                windowed_pos_all,
                groupsToPlotFst,
                groupColorsFst;
                lineTransparency = 0.8,
                fillTransparency = 0.2,
                figureSize = (1200, 1200))</code></pre><p>Plot windowed Fst along chromosomal regions, for multiple population comparisons.</p><p>​# Arguments</p><ul><li><code>scaffolds_to_plot</code>: a vector of the scaffold names to include in the plot.</li><li><code>windowed_Fst_all</code>: a matrix of Fst values (rows representing different population comparisons; columns different genomic regions).</li><li><code>pairwiseNamesFst</code>: a vector of names indicating the pairs of populations in rows of <code>windowed_Fst_all</code>.</li><li><code>windowed_pos_all</code>: dataframe containing genomic locations of windows (must have columns <code>chrom</code> and <code>mean_position</code>).</li><li><code>groupsToPlotFst</code>: a vector of strings indicating the population comparisons to include.</li><li><code>groupColorsFst</code>: the colors to use to plot the comparisons in <code>groupsToPlotFst</code>.</li><li><code>lineTransparency</code> (optional): transparency of the lines.</li><li><code>fillTransparency</code> (optional): transparency of the fill color.</li><li><code>figureSize</code> (optional): size of the figure (height, width).</li></ul><p><strong>Notes</strong></p><ul><li>Returns the figure.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L1291-L1317">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.plotGenomeViSHet-Tuple{Any, Any, Any}" href="#GenomicDiversity.plotGenomeViSHet-Tuple{Any, Any, Any}"><code>GenomicDiversity.plotGenomeViSHet</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plotGenomeViSHet(scaffolds_to_plot, 
                windowed_ViSHet_all,
                windowed_pos_all;
                fillColor = &quot;purple&quot;,
                lineTransparency = 0.8,
                fillTransparency = 0.2,
                figureSize = (1200, 1200),
                plotRegions = false,
                regionsToPlot = [],
                regionColor = &quot;red&quot;)</code></pre><p>Plot windowed Variance in Standardized Heterozygosity (ViSHet) along multiple scaffolds.</p><p>​# Arguments</p><ul><li><code>scaffolds_to_plot</code>: a vector of the scaffold names to include in the plot.</li><li><code>windowed_ViSHet_all</code>: a vector of ViSHet values for windows of the genome.</li><li><code>windowed_pos_all</code>: dataframe containing genomic locations of windows (must have columns <code>chrom</code> and <code>mean_position</code>).</li><li><code>fillColor</code> (optional): the color for the lines and fill (default <code>&quot;purple&quot;</code>).</li><li><code>lineTransparency</code> (optional): transparency of the lines.</li><li><code>fillTransparency</code> (optional): transparency of the fill color.</li><li><code>figureSize</code> (optional): size of the figure (height, width).</li><li><code>plotRegions</code> (optional): choose <code>true</code> to highlight regions.</li><li><code>regionsToPlot</code> (optional): a DataFrame with three columns: <code>regionChrom</code> containing names of scaffolds, <code>regionStart</code> and <code>regionEnd</code> giving start and end of regions to highlight.</li><li><code>regionColor</code> (optional): the color for the highlighted regions.</li></ul><p><strong>Notes</strong></p><ul><li>Returns the figure.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L1495-L1523">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.plotGenotypeByIndividual-NTuple{7, Any}" href="#GenomicDiversity.plotGenotypeByIndividual-NTuple{7, Any}"><code>GenomicDiversity.plotGenotypeByIndividual</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plotGenotypeByIndividual(regionInfo, pos, genoData, 
                            indMetadata, freqs,
                            plotGroups, plotGroupColors;
                            missingFractionAllowed = 1.0,
                            colorAllelesByGroup = true, group1 = plotGroups[1],
                            indFontSize=10, figureSize=(1200, 1200),
                            show_SNP_density = true, densityPlotColor = :steelblue1,
                            plotTitle = nothing, titleFontSize = 20,
                            indColorLeftProvided = false,
                            indColorRightProvided = false)</code></pre><p>Construct a genotype-by-individual plot, with option to filter out SNPs with too much missing data. </p><p>Under the default setting, alleles are colored (dark purple vs. light purple) according to whichever allele is designated as <code>group1</code>. </p><p>​# Arguments</p><ul><li><code>regionInfo</code>: Information regarding the scaffold and region of focus; a tuple provided by <code>chooseChrRegion()</code>.</li><li><code>pos</code>: Matrix providing genomic location of each locus; there must be a <code>chrom</code> column and a <code>position</code> column.</li><li><code>genoData</code>: Matrix containing genotype data (individuals in rows, loci in columns).</li><li><code>indMetadata</code>: Matrix of metadata for individuals; must contain <code>Fst_group</code> and <code>plot_order</code> columns.</li><li><code>freqs</code>: Matrix of alternate allele frequencies for each group (row) and locus (column).</li><li><code>plotGroups</code>: Vector of group names to include in plot.</li><li><code>plotGroupColors</code>: Vector of plotting colors corresponding to the groups.</li><li><code>missingFractionAllowed</code>: Optional; The maximum missing genotype fraction for a locus to be included (default is 1.0, so no filtering).</li><li><code>colorAllelesByGroup</code>: Optional; set to <code>false</code> to color alleles according to reference and alternate.</li><li><code>group1</code>: Optional (default is <code>plotGroups[1]</code>); when <code>colorAllelesByGroup</code> is <code>true</code>, this is the group that determine which allele is dark purple. </li><li><code>indFontSize</code>: Optional; the font size of the individual ID labels.</li><li><code>figureSize</code>: Optional; the size of the figure; default is <code>(1200, 1200)</code>.  </li><li><code>show_SNP_density</code>: Optional; default is <code>true</code> to show a density plot. </li><li><code>densityPlotColor</code>: Optional; default is <code>:steelblue1</code></li><li><code>plotTitle</code>: Optional; default will make a title. For no title, set to <code>&quot;&quot;</code>.</li><li><code>titleFontSize</code>; Optional; default is 20.</li><li><code>indColorLeftProvided</code>: Optional; Default is <code>false</code>. Set to <code>true</code> if there is a column labeled <code>indColorLeft</code> in the metadata providing color of each individual for plotting on left side.</li><li><code>indColorRightProvided</code>: Optional; same as above but for right side (requires <code>indColorRight</code> column in metadata).</li></ul><p><strong>Notes</strong></p><p>Returns a tuple containing:</p><ul><li>the figure</li><li>the plotted genotypes</li><li>the numerical positions (in the chosen scaffold) of the plotted loci</li><li>the sorted metadata matrix for the plotted individuals</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L719-L762">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.plotGenotypeByIndividualWithFst-NTuple{12, Any}" href="#GenomicDiversity.plotGenotypeByIndividualWithFst-NTuple{12, Any}"><code>GenomicDiversity.plotGenotypeByIndividualWithFst</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plotGenotypeByIndividualWithFst(groupsToCompare, Fst_cutoff, missingFractionAllowed,
                        regionInfo, pos, Fst, pairwiseNamesFst,
                        genoData, indMetadata, freqs, plotGroups, plotGroupColors;
                        colorAllelesByGroup = true, group1 = plotGroups[1],
                        indFontSize=10, figureSize=(1200, 1200),
                        show_SNP_density = true, densityPlotColor = :steelblue1,
                        plotTitle = nothing, titleFontSize = 20,
                        highlightRegionStarts = [],
                        highlightRegionEnds = [],
                        highlightRegionColor = &quot;red&quot;)</code></pre><p>Construct a genotype-by-individual plot, including only loci that pass thresholds for Fst and missing data. </p><p>Under the default setting, alleles are colored (dark purple vs. light purple) according to whichever allele is designated as <code>group1</code>. </p><p>​# Arguments</p><ul><li><code>groupsToCompare</code>: The two groups (in format <code>name1_name2</code>) for filtering loci based on Fst. This can be a single string, or a vector of strings, each with a population comparison to include if Fst is above Fst_cutoff.</li><li><code>Fst_cutoff</code>: The minimum Fst for a locus to be included in the plot.</li><li><code>missingFractionAllowed</code>: The maximum missing genotype fraction for a locus to be included.</li><li><code>regionInfo</code>: Information regarding the scaffold and region of focus; a tuple provided by <code>chooseChrRegion()</code>.</li><li><code>pos</code>: Matrix providing genomic location of each locus; there must be a <code>chrom</code> column and a <code>position</code> column.</li><li><code>Fst</code>: Matrix of Fst values; can be produced by <code>getFst()</code>.</li><li><code>pairwiseNamesFst</code>: Vector of pairwise names corresponding to rows in the <code>Fst</code> matrix.</li><li><code>genoData</code>: Matrix containing genotype data (individuals in rows, loci in columns).</li><li><code>indMetadata</code>: Matrix of metadata for individuals; must contain <code>Fst_group</code> and <code>plot_order</code> columns.</li><li><code>freqs</code>: Matrix of alternate allele frequencies for each group (row) and locus (column).</li><li><code>plotGroups</code>: Vector of group names to include in plot.</li><li><code>plotGroupColors</code>: Vector of plotting colors corresponding to the groups.</li><li><code>colorAllelesByGroup</code>: Optional; set to <code>false</code> to color alleles according to reference and alternate.</li><li><code>group1</code>: Optional (default is <code>plotGroups[1]</code>); when <code>colorAllelesByGroup</code> is <code>true</code>, this is the group that determine which allele is dark purple. </li><li><code>indFontSize</code>: Optional; the font size of the individual ID labels.</li><li><code>figureSize</code>: Optional; the size of the figure; default is <code>(1200, 1200)</code>.  </li><li><code>show_SNP_density</code>: Optional; default is <code>true</code> to show a density plot. </li><li><code>densityPlotColor</code>: Optional; default is <code>:steelblue1</code></li><li><code>plotTitle</code>: Optional; default will make a title. For no title, set to <code>&quot;&quot;</code>.</li><li><code>titleFontSize</code>; Optional; default is <code>20</code>.</li><li><code>highlightRegionStarts</code>: Optional; the left locations of regions to highlight on the scaffold.</li><li><code>highlightRegionEnds</code>: Optional; the right locations of regions to highlight on the scaffold. </li><li><code>highlightRegionColor</code>: Optional; default red; the color of the highlight bar along the scaffold.</li></ul><p><strong>Notes</strong></p><p>Returns a tuple containing:</p><ul><li>the figure</li><li>the plotted genotypes</li><li>the numerical positions (in the chosen scaffold) of the plotted loci</li><li>the sorted metadata matrix for the plotted individuals</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L945-L992">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.plotGenotypeByIndividualWithFst-Tuple{GenoData, Vararg{Any, 9}}" href="#GenomicDiversity.plotGenotypeByIndividualWithFst-Tuple{GenoData, Vararg{Any, 9}}"><code>GenomicDiversity.plotGenotypeByIndividualWithFst</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plotGenotypeByIndividualWithFst(myGenoData::GenoData, groupsToCompare, 
                        Fst_cutoff, missingFractionAllowed,
                        regionInfo, Fst, pairwiseNamesFst,
                        freqs, plotGroups, plotGroupColors;
                        colorAllelesByGroup = true, group1 = plotGroups[1],
                        indFontSize=10, figureSize=(1200, 1200),
                        show_SNP_density = true, densityPlotColor = :steelblue1,
                        plotTitle = nothing, titleFontSize = 20,
                        highlightRegionStarts = [],
                        highlightRegionEnds = [],
                        highlightRegionColor = &quot;red&quot;)</code></pre><p>Call the plotGenotypeByIndividualWithFst() function using a <code>GenoData</code> object.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L1191-L1205">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.plotPCA-NTuple{4, Any}" href="#GenomicDiversity.plotPCA-NTuple{4, Any}"><code>GenomicDiversity.plotPCA</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plotPCA(genotypes, ind_with_metadata, groups_to_plot_PCA, group_colors_PCA; sampleSet = &quot;&quot;, regionText=&quot;&quot;, flip1 = false, flip2 = false)</code></pre><p>Do a principal components analysis based on genotypes of individuals (colored by group), and display PC1 vs. PC2.</p><p>​# Arguments</p><ul><li><code>genotypes</code>: Genotype matrix (individuals in rows, loci in columns, with no missing data–can impute prior to this, to ensure no missing).</li><li><code>indMetadata</code>: Matrix containing metadata for individuals (make sure there is an <code>Fst_group</code>` column).</li><li><code>groups_to_plot_PCA</code>: Vector of names of groups to include.</li><li><code>group_colors_PCA</code>: Vector listing plotting color for each each group.</li></ul><p>Optional arguments:</p><ul><li><code>sampleSet</code>: Name of the sample set to appear in the plot title.</li><li><code>regionText</code>: Name of the genomic region to appear in the plot title.</li><li><code>flip1</code>: Set to <code>true</code> if wanting to flip PC1 (i.e., multiply by -1).</li><li><code>flip2</code>: Same but for PC2.</li><li><code>showPlot</code>: Set to <code>false</code> if not wanting the function to draw display.</li><li><code>autolimitaspect_setting</code>: For unconstrained axes, set to <code>nothing</code> (default is <code>1</code> for 1:1 axes).</li><li><code>lineOpacity</code>: Opacity of the line color of symbols.</li><li><code>fillOpacity</code>: Opacity of the fill color of symbols.</li><li><code>symbolSize</code>: Size of symbols.</li><li><code>plotTitle</code>: Specify if a specific title is wanted; otherwise will fill in a title.</li><li><code>showTitle</code>: Set to <code>false</code> for no title. </li><li><code>xLabelText</code>: Can choose an x axis label.</li><li><code>yLabelText</code>: Can choose a y axis label.</li><li><code>labelSize</code>: Size of x and y axis labels.</li></ul><p><strong>Notes</strong></p><p>Returns a tuple containing:</p><ol><li><code>model</code>: the PCA model.</li><li><code>metadata</code>: the metadata of individuals included in the PCA.</li><li><code>values</code>: PC values of individuals (before any flipping of axes).</li><li><code>PC1</code>: PC1 values of individuals(after any flipping).</li><li><code>PC2</code>: PC2 values of individuals (after any flipping).</li><li><code>PCAfig</code>: The PCA figure.</li></ol></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L529-L564">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.plotPCA-Tuple{GenoData, Any, Any}" href="#GenomicDiversity.plotPCA-Tuple{GenoData, Any, Any}"><code>GenomicDiversity.plotPCA</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plotPCA(myGenoData::GenoData, groups_to_plot_PCA, group_colors_PCA; sampleSet = &quot;&quot;, regionText=&quot;&quot;, flip1 = false, flip2 = false)</code></pre><p>Apply the plotPCA function to a <code>GenoData</code>` object.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L646-L650">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicDiversity.standardizeIndHet-Tuple{Any}" href="#GenomicDiversity.standardizeIndHet-Tuple{Any}"><code>GenomicDiversity.standardizeIndHet</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">standardizeIndHet(windowedIndHet)</code></pre><p>Divide each individual&#39;s windowed heterozgyosity by the average heterozygosity of all individuals (at each window).</p><p>​# Arguments</p><ul><li><code>windowedIndHet</code>: The matrix of individual windowed heterozygosity values (rows are individuals; columns represent windowed genomic regions)</li></ul><p><strong>Notes</strong></p><p>Returns a matrix of standardized individual windowed heterozygosities.  </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/darreni/GenomicDiversity.jl/blob/12ef2418a94d096ad0fd209f51b8a196d71b57af/src/GenomicDiversity.jl#L1433-L1443">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.1 on <span class="colophon-date" title="Thursday 13 March 2025 21:58">Thursday 13 March 2025</span>. Using Julia version 1.11.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
